File: php/features.php
<?php
declare(strict_types=1);

// OpenGovernmentPortal SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class OpenGovernmentPortalFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new OpenGovernmentPortalBaseFeature();
            case "ratelimit":
                return new OpenGovernmentPortalRatelimitFeature();
            case "retry":
                return new OpenGovernmentPortalRetryFeature();
            case "test":
                return new OpenGovernmentPortalTestFeature();
            case "timeout":
                return new OpenGovernmentPortalTimeoutFeature();
            default:
                return new OpenGovernmentPortalBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: py/opengovernmentportal_sdk/features.py
# OpenGovernmentPortal SDK feature factory

from opengovernmentportal_sdk.feature.base_feature import OpenGovernmentPortalBaseFeature
from opengovernmentportal_sdk.feature.ratelimit_feature import OpenGovernmentPortalRatelimitFeature
from opengovernmentportal_sdk.feature.retry_feature import OpenGovernmentPortalRetryFeature
from opengovernmentportal_sdk.feature.test_feature import OpenGovernmentPortalTestFeature
from opengovernmentportal_sdk.feature.timeout_feature import OpenGovernmentPortalTimeoutFeature


_FEATURES = {
    "base": lambda: OpenGovernmentPortalBaseFeature(),
    "ratelimit": lambda: OpenGovernmentPortalRatelimitFeature(),
    "retry": lambda: OpenGovernmentPortalRetryFeature(),
    "test": lambda: OpenGovernmentPortalTestFeature(),
    "timeout": lambda: OpenGovernmentPortalTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
